### PR TITLE
Set ownership of uploaded files to correct user

### PR DESCRIPTION
**Pull request by jcapona**
| Status  | Ticket/Issue |
| :---: | :--: |
| Ready/Hold | [Ticket](https://pi-top.atlassian.net/browse/OS-1265) |

#### Main changes
-

#### Screenshots (feature, test output, profiling, dev tools etc)

[insert screenshots here]

#### Other notes (e.g. implementation quirks, edge cases, questions / issues)
-

#### Manual testing tips
-

#### Tag anyone who definitely needs to review or help
-

### CODE REVIEW

File: further_link/util/upload.py [hunk 1]
```diff
@@ -32,6 +34,25 @@ def directory_is_valid(directory):
...
+    splitted_path = directory_path.split("/")
+    subpaths = ["/".join(splitted_path[:i]) for i in range(2, len(splitted_path) + 1)]
+
+    for subpath in subpaths:
+        if not os.path.isdir(subpath):
+            os.mkdir(subpath)
+
+            # Update directory owner if on user home
+            if user and subpath.startswith(os.path.expanduser(f"~{user}")):
+                os.chown(subpath, uid=get_uid(user), gid=get_gid(user))
```
Comment (angusjfw) [inline further_link/util/upload.py:53]: it's a shame this doesn't have a recursive option, what a hassle!

File: further_link/util/upload.py [hunk 2]
```diff
@@ -32,6 +34,25 @@ def directory_is_valid(directory):
     )
 
 
+def create_directory(directory_path: str, user: str = None):
+    """
+    Create the directories from the provided path level by level, making sure the folders have the correct owner
+    """
+    if user is None:
+        user = get_user_using_first_display()
```
Comment (angusjfw) [inline further_link/util/upload.py:42]: in other places i'm using `utils.user_config.default_user`, which basically uses hardcoded 'pi' user if it's available. this is better, so I think we should change all those places to use `utils.sdk.get_user_using_first_display` for consistency please

File: further_link/runner/exec_process_handler.py [hunk 3]
```diff
@@ -15,7 +16,7 @@ async def start(self, path, code=None, novncOptions={}):
         path = get_absolute_path(path, get_working_directory(self.user))
 
         # create path directories if they don't already exist
-        os.makedirs(os.path.dirname(path), exist_ok=True)
+        create_directory(os.path.dirname(path), self.user)
 
         entrypoint = path if code is None else os.path.join(path, f"exec-{self.id}")
```
Comment (angusjfw) [inline further_link/runner/exec_process_handler.py:22]: right after this we (optionally) write a file with the user code. even though in that case the file is temporary, maybe we should still chown it to the user? there could be cases where it matters 🤔

**Review by angusjfw — COMMENTED** [review]
spot on! i just have a couple extra suggestions to keep things consistent.

File: further_link/runner/exec_process_handler.py [hunk 4]
```diff
@@ -25,6 +31,7 @@ async def start(self, path, code=None, novncOptions={}):
                 await file.write(code)
             self._remove_entrypoint = entrypoint
 
+        os.chown(entrypoint, uid=get_uid(self.user), gid=get_gid(self.user))
```
Comment (angusjfw) [inline further_link/runner/exec_process_handler.py:34]: we should do this in shell_process_handler and py_process_handler too. run_py i care less about as that is part of a deprecated API

Comment (jcapona) [reply further_link/runner/exec_process_handler.py:34]: didn't find anything in `shell_process_handler`... just updated `py_process_handler` & `run_py`
